[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47796 [PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()**

`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

swolchok committed:
[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()

`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
add missing brace on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
fix build on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
s/getRawUnsafe/get on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
fix build on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
clang-format on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

ezyang commented:
I can review this in a pinch but I'm not the best person to look at the memory profiling bits. Give a holler if you do want me to fully review this.

dzhulgakov commented on c10/util/ThreadLocalDebugInfo.h:
```diff
@@ -32,7 +32,7 @@ class C10_API DebugInfoBase {
 class C10_API ThreadLocalDebugInfo {
  public:
-  static std::shared_ptr<DebugInfoBase> get(DebugInfoKind kind);
+  static DebugInfoBase* get(DebugInfoKind kind);
```
Maybe add a comment about lifetime of the returned object?

dzhulgakov reviewed (approved):
I think it's good to ship

swolchok committed:
Update on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]

swolchok committed:
clang-format on "[PyTorch] Add ThreadLocalDebugInfo::getRawUnsafe & use to replace get()"


`ThreadLocalDebugInfo::get()` is a hot function. For example, it is called by `DefaultCPUAllocator::allocate()`. Most callers do not even bother to keep the returned `shared_ptr` around, proving that they have no lifetime issues currently. For the rest, it appears that the only way that the returned pointer could become invalid is if they then called a function that swapped out `ThreadLocalDebugInfo` using `ThreadLocalStateGuard`. There are very few such paths, and it doesn't look like any current callers of `ThreadLocalDebugInfo::get()` needed a `shared_ptr` at all.

Differential Revision: [D24902978](https://our.internmc.facebook.com/intern/diff/D24902978/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24902978/)!

[ghstack-poisoned]
